[Feature] Add custom user-defined game categories #1428

lgcavalheiro opened the pull request:
<--- Put the description here --->

This should close or partially satisfy https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/issues/1428
This is a very simple system for creating custom categories and adding/removing games to them, similar to how hidden and favorites work.

On the category settings modal, when a category is assigned to the game, clicking on the "X" will remove that game from the category, when a category is in the `available categories` section, clicking the body of the chip will assign the game to the category, clicking on the "X" will delete the category.

On the library view there's now a dropdown selector to filter games by a category, this filter also stacks with the hidden and platform filters.

Video demonstration:
https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/10094385/856f5575-02a8-49cc-a727-b2b4df4b2f30

Sample of `store/config.json`:
![image](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/10094385/b9d5c8a6-ae7c-416b-8144-980f529dfc11)

---

Use the following Checklist if you have changed something on the Backend or Frontend:

- [x] Tested the feature and it's working on a current and clean install.
- [x] Tested the main App features and they are still working on a current and clean install. (Login, Install, Play, Uninstall, Move games, etc.)
- [ ] Created / Updated Tests (If necessary)
- [ ] Created / Updated documentation (If necessary)

arielj commented on public/locales/es/gamepage.json:
```diff
@@ -231,7 +231,8 @@
         "extraInfo": "Información extra",
-        "logs": "Registros"
+        "logs": "Registros",
+        "manageCategories": "Administrar Categorías"
```
we don't add translations for languages other than english here in the code, the spanish translations should be later added to weblate once this PR is merge

this file can be overrode by weblate if we only have change here

arielj commented:
I'm worried about the `x` in the `available categories` categories deleting the category completely. It's too easy to delete things by mistake (maybe we should display a warning saying "This category is applied to these games: ...., are you sure you want to delete it?").

arielj commented on src/frontend/types.ts:
```diff
@@ -55,12 +55,22 @@ export interface ContextType {
+  customCategories: {
+    list: Record<string, string[]>
+    listCategories: () => string[]
+    add: (category: string, appName: string) => void
+    remove: (category: string, appName: string) => void
```
```suggestion
    addToGame: (category: string, appName: string) => void
    removeFromGame: (category: string, appName: string) => void
```

so it's more consistent with the next 2 functions

arielj commented on src/frontend/screens/Library/components/GameCard/index.tsx:
```diff
@@ -338,6 +338,11 @@ const GameCard = ({
       show: !isFavouriteGame
     },
+    {
+      label: t('submenu.manageCategories', 'Manage Categories'),
```
```suggestion
      label: t('submenu.categories', 'Categories'),
```

I think it's still clear and makes it easier for translators (in some languages maybe it's long to say "Manage Categories" and then the card menu might look weird)

arielj reviewed (commented):
We probably want to add a button in the 3 dots menu in the game page to open the same Categories dialog

lgcavalheiro commented:
Fixed issues mentioned above, now before deleting a category outright, a confirmation dialog will show up:
![image](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/10094385/d250f629-22a3-4750-95bd-2ddf281ebb09)

Another idea i had was to implement an undo pattern instead (example from https://goodui.org):
![image](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/10094385/4d56ef76-d1bd-4544-be16-44eff99b959c)

I've also took the liberty of adding a preset category (uncategorized) which will list all games that have no categories assigned to them.

imLinguin commented:
Could you also add runner for each game? To avoid issues when the same appid is found in another store, as we plan to extend Heroic to more integrations eventually

arielj commented:
I wonder if it's easier to use if we use checkboxes instead of pills. It will also make the UI less "jumpy".

One thing I was expecting was that if I add a new category, it should set that already for the game I'm configuring. Currently I have to first add a category, then it appears in the available ones, then I have to click it.

One thing I noticed is that you can type the same category twice when adding a category, it lets you submit the category but it unassigns every game from it. I think we should disable the submit button when the category typed there already exists.

One error I found is that you can delete a category while you are filtering by it, when you close the modal the app crashes with:
![image](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/188464/ee9af182-432d-4df5-bcf4-35483b054dfe)

I think we should also trim the category name, you can add `"           "` with any number of white spaces, or `"       my category      "` for example

arielj commented:
Something weird happens with the focus when I use a mouse to control the UI, when I focus the category it goes transparent so I can't see the content, pressing `A` adds the category if it's in the Available column, pressing A does nothing if it's in the Assigned column.

Pressing A does nothing when focusing the `X` button

https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/188464/a8f749a5-3ea7-4d19-bac4-2fe624b06099

Maybe using checkboxes and labels fixes that? we already have the toggle component that we use in the settings and it's controller friendlier. We could have:

```
[  ]  Category 1                        ( x )
[  ]  Category 2                        ( x )
```

flavioislima commented:
@lgcavalheiro what is the current state of this PR? are you gonna fix the other comments or should we assume it? I think there is just a few small things to improve and one conflict to fix as well :)

lgcavalheiro commented:
@flavioislima sry for the hiatus, got busy with work and college these past few weeks, i'll be resuming work on this PR this week.

flavioislima commented:
> @flavioislima sry for the hiatus, got busy with work and college these past few weeks, i'll be resuming work on this PR this week.

No problem! 👍

lgcavalheiro commented:
Hi folks, fixed the issues mentioned, here's how this feature is looking at the moment:

[finalforpr_2.webm](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/10094385/248e6a50-ea61-4f9d-848c-2c03b19011fb)

- Using appName and runner as well
- Limited category length to 33 (seems to be the limit on steam, so i went with that)
- Trims leading and trailling white spaces from category names
- Does not allow for duplicate categories anymore
- If deleting a category that is being used as filter, will reset filter before deleting the category
- When a new category is created, it will be assigned to the current game as well
- Changed layout in favor of checkboxes, controller testing seems ok

Exemple of new customCategories in store/config.json:
``` json
"customCategories": {
			"2D": [
				"1dea8a6ddb544842a58e4b5c8675ff58_legendary",
				"amzn1.adg.product.4a10b5b0-d498-4a74-8053-e31f665dc82c_nile"
			],
			"Action": [
				"1dea8a6ddb544842a58e4b5c8675ff58_legendary",
				"amzn1.adg.product.4a10b5b0-d498-4a74-8053-e31f665dc82c_nile",
				"Petunia_legendary",
				"1133514031_gog"
			]
}
```

flavioislima commented:
@lgcavalheiro those are really nice improvements, thanks for that.

Latest issues are just related to layout on small windows, like this:
![image](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/26871415/8475ccd5-802c-4e0c-8cea-cd07d10a9bac)

Maybe creating another row could fix that. we should have a media query for that.

lgcavalheiro commented:
@flavioislima got a basic layout fix going for smaller screens, see if it's better now (didn´t really know how much i should pay attention to responsiveness, since it's a desktop application and stuff, sry):

[finalforpr_3_trimmed.webm](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/10094385/be9e2a20-c827-4589-8156-a7d4e26d4e7a)

arielj commented:
In this PR https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/pull/3266 I'm moving all the filter into a `Filters` button, that will save some space there (then we could put the categories selector either next to it instead of its own row, or even inside the filters maybe?).

I also think we have to make some decisions:
- what's the minimum width we want to support? (so we know when to stop)
- should we auto-collapse the sidebar when the width is too small? that will free some room for the library
(of course these questions are unrelated to this PR but something we should do for the whole project, cc @flavioislima )


What I mean is that maybe we can improve the small screens design when both things are merged and those decisions are made.

lgcavalheiro commented:
@arielj I did a quick local merge of your PR with mine and having the categories selector by the filters button made the UI look much cleaner indeed :D 
I suppose i'll just wait until the new filters menu is rolled out, then i'll rebase against upstream before you guys roll the custom categories out.

![image](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/10094385/c3c400e0-31ab-42e6-9c43-b7ab299f03b5)

![image](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/10094385/612fdcf7-79b7-45a6-b517-6b5e47d96e8b)

About supported widths, my 2 cents would be supporting a minimum between 758px (lowest width i could observe in which Heroic still looks good and nothing is visibly breaking) and 960px (half width of a FHD monitor)

arielj commented:
@lgcavalheiro my PR was just merged if you want to update this one against `main`

lgcavalheiro commented:
Hi guys i just finished merging and manually testing stuff to be sure everything is working ok, the merge process was a bit messy so i'd recommend double-checking just to be sure @arielj @flavioislima 

On a side note: do you think the search bar suggestions should take the filters into consideration when searching for games? Right now it is not:

[finalforpr_4.webm](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/assets/10094385/48614a83-dfc1-4940-9ca6-a324dba2b526)

flavioislima commented:
> 
> On a side note: do you think the search bar suggestions should take the filters into consideration when searching for games? Right now it is not:

That is a pretty good question. 
I think it should not because you can now click on the selected item on the search bar and it will redirect you to the game page. So if we limit to the current filters this won't be possible. So I think we should keep it as it is.

I imagine the following scenario, for instance. 
A person enabled some filters and cannot find a certain game. It decides to use the search bar. Right now it will find the game, clicking on it will go to the game page. If we change that, the game will never be found. So I think it is good to have all games there. 🙂

flavioislima reviewed (approved):
Tested everything here and it is working pretty fine.
Thanks a lot for adding this feature to Heroic and I believe it is safe to be on the next release ⚔️ 🛡️
